### DISCUSSION TITLE
"using(Py.GIL())" must be called on same thread as PythonEngine.Initialize()?

### DISCUSSION THREAD

**Discussion by andyste1** [pythonnet/pythonnet#2075] [Q&A]
We have a WPF desktop data processing app (.Net 6, Win10), and part of the processing pipeline involves executing Python code (often many hundreds of times in succession). We've been using IronPython to do this but customers are asking to use libraries such as numpy, so I've refactored the app to replace IP with Python.Net v3.0.1 (Python 3.8).

I call PythonEngine.Initialize() once only, during application startup, and .Shutdown() when the app closes, both of which run on the UI thread. There are numerous places throughout the app's code that executes Python code, some of which run on the UI thread but most are wrapped in Task.Run() so run on b/g threads. During testing I have discovered that the Python execution code ("using(Py.GIL())" block) must run on the same thread as what PythonEngine.Initialize() was run on (the UI thread in my case). If it runs on a different thread then the call to Py.GIL() never returns and hangs the app.

My options seem to be:

- Call PythonEngine.Initialize() before the using(Py.GIL()) block (and followed by Shutdown() of course), ensuring everything runs on the same thread. This does fix the issue but adds an unacceptable overhead of several hundred milliseconds. (The code within the "using" block typically only takes a few tens of ms to execute).

- Somehow ensure that Initialize(), the "using" block, and Shutdown(), always run on the same thread. I have come up with a solution using the .Net BlockingCollection class, but it feels very hacky and I can't help wondering if if there's a better solution, or something that I'm missing in how to use Python.Net/Python.

**Comment by filmor** [comment]
Python's main thread will hold the GIL until you explicitly release it, see https://github.com/pythonnet/pythonnet/issues/109. We should probably add this to the docs (https://github.com/pythonnet/pythonnet/blob/master/doc/source/dotnet.rst), a PR would be appreciated :)

**Reply by filmor** [reply]
Actually, it is already documented here: https://github.com/pythonnet/pythonnet/wiki/Threading, we just need to move that to the central location (like some of the other wiki pages).

**Reply by andyste1** [reply]
Thanks for the quick reply, that did the trick. 

Just to clarify a related question: my app runs certain data processing tasks in parallel (e.g. Parallel.ForEach), and those tasks might involve executing Python code. Am I right in saying that this will be fine as long as I follow the advice in that wiki article? 
In this situation, am I right in saying that only one thread will be able to execute Python code at a time, with the "using(Py.GIL())" construct effectively acting as a lock?

**Reply by filmor** [reply]
Yes, that is right. The L in GIL means Lock, after all ;). However, if your threads "drop down" to C, e.g. if your data processing uses NumPy or if they are waiting for IO, they will release the GIL for that part and thus run (or wait) in parallel.

**Reply by stephen-swensen** [reply]
In this scenario will the python interpreter ever pre-emptively drop the GIL in pure python code? i.e. suppose using `Parallel.ForEach` and each C# thread executes some dynamic python script like

```
using (PyModule scope = Py.CreateScope())
{
   //...import libraries, set vars
   scope.Exec(script); //complex, long running script with lots of statements
   //...get vars
}
```

will time be spread out somewhat evenly executing each python script or will they be run strictly in series. Realistic scenario is a web server with many concurrent requests each executing some python script. It would be OK for each request to be equally slow, but less OK if one unlucky request has to wait for all the others to complete before it can get its result.
